src/Nethermind/Nethermind.State/Proofs/ReceiptTrie.cs

```diff
@@ -7,29 +7,28 @@
 using Nethermind.Core;
 using Nethermind.Core.Buffers;
 using Nethermind.Core.Crypto;
-using Nethermind.Core.Extensions;
 using Nethermind.Core.Specs;
 using Nethermind.Serialization.Rlp;
 using Nethermind.State.Trie;
 using Nethermind.Trie;
-using Nethermind.Trie.Pruning;
 
 namespace Nethermind.State.Proofs;
 
 /// <summary>
 /// Represents a Patricia trie built of a collection of <see cref="TxReceipt"/>.
 /// </summary>
-public class ReceiptTrie : PatriciaTrie<TxReceipt>
+public class ReceiptTrie<TReceipt> : PatriciaTrie<TReceipt>
 {
-    private static readonly ReceiptMessageDecoder _decoder = new();
-
+    private readonly IRlpStreamDecoder<TReceipt> _decoder;
     /// <inheritdoc/>
     /// <param name="receipts">The transaction receipts to build the trie of.</param>
-    public ReceiptTrie(IReceiptSpec spec, IEnumerable<TxReceipt> receipts, bool canBuildProof = false, ICappedArrayPool? bufferPool = null)
+    public ReceiptTrie(IReceiptSpec spec, IEnumerable<TReceipt> receipts, IRlpStreamDecoder<TReceipt> decoder, bool canBuildProof = false, ICappedArrayPool? bufferPool = null)
```

Comment:
The only thing we care about is the Decoder, if we pass one that takes base TxReceipt here we might avoid making this class generic? Not sure if good idea.

Reply:
I made it that way because we might have completely new type of receipt. And in addition generics >> interfaces. Because you can't pass wrong decoder accidentally.